File: src/implementation/PrivateChatManager.js
class PrivateChatManager {
    constructor(configManager, telegramBot, calendarManager) {
        this.configManager = configManager;
        this.telegramBot = telegramBot;
        this.calendarManager = calendarManager;
        this.pullOffset = configManager.getProperty('PULL_OFFSET') ? parseInt(configManager.getProperty('PULL_OFFSET')) : 0;

    }

    setupMenu() {
        const commands = [
            { command: "register", description: "Зареєструватися" },
            { command: "book", description: "Забронювати" },
            { command: "help", description: "Допомога" }
        ];
        this.telegramBot.setMyCommands(commands);
    }

    pollUpdates(timeoutSeconds = 5) {
        const payload = {
            offset: this.pullOffset,
            limit: 100,
            timeout: timeoutSeconds
        };
        const updates = this.telegramBot.getUpdatesApi(payload);

        if (Array.isArray(updates) === false) {
            Utils.logError(`Error fetching updates: Not an array: ${updates}`);
            return;
        }

        if (!updates || updates.length === 0) {
            Utils.logInfo("No updates received or error occurred.");
            return;
        }

        try {
            this.processUpdates(updates);
        } catch (error) {
            Utils.logError(`Error processing updates: ${error}`);
            return;
        }
        finally {
            this.pullOffset = updates[updates.length - 1].update_id + 1;
            this.configManager.setProperty('PULL_OFFSET', this.pullOffset.toString());
        }
    }

    processUpdates(updates) {
        updates.forEach(update => {

            try {
                var updateLogData = JSON.stringify(update)

                if (update?.message?.chat?.type !== 'private' &&
                    update?.callback_query?.message?.chat?.type !== 'private') {

                    // Utils.logInfo(`Skipping undefined update ${updateLogData}`);
                    return;
                }

                // TODO remove to mamy logs. Only for testing
                Utils.logInfo(`Received update ${updateLogData}`);

                if (update.message) {
                    const message = update.message;
                    if (message.text === undefined) {
                        Utils.logInfo(`Received message without text from ${message.from.username}`);
                        return;
                    }

                    Utils.logInfo(`Received message: ${message.text} from ${message.from.username}`);
                    if (message.text.startsWith('/')) {
                        this.handleCommand(update);
                    }
                    if (message.text.includes('@')) {
                        this.handleEmailMessage(message);
                    }
                } else if (update.callback_query) {
                    const callbackQuery = update.callback_query;
                    Utils.logInfo(`Received callback query: ${callbackQuery.data} from ${callbackQuery.from.username}`);
                    this.handleCallbackQuery(callbackQuery);
                }
            } catch (error) {
                Utils.logError(`Error processing update: Error: ${error}. Update: ${JSON.stringify(update)}`);
            }
        });
    }

    handleEmailMessage(message) {
        const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|.(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        if (message.text.match(EMAIL_REGEX)) {
            Utils.logInfo(`Email detected: ${message.text}`);
            // TODO: Handle email registration
            this.telegramBot.sendMessage(
                message.chat.id,
                `Ваш емейл ${message.text} успішно зареєстровано. Дякуємо!`
            );
        }
    }

    handleCommand(update) {
        switch (update.message.text) {
            case '/book':
                Utils.logInfo("Handling book command");
                this.telegramBot.sendMessage(
                    update.message.chat.id,
                    "Оберіть дату та час для бронювання:",
                    { inline_keyboard: this.getAvailableDates(null) }
                );
                break;
            case '/help':
            case '/start':
                Utils.logInfo("Handling help command");
                this.telegramBot.sendMessage(
                    update.message.chat.id,
                    "Довідка: використовуйте /book для бронювання.\nДля реєстрації/оновлення емейла використовуйте /register."
                );

                // TODO mark that user is writing email

                break;
            case '/register':
                Utils.logInfo("Handling register command");
                this.telegramBot.sendMessage(
                    update.message.chat.id,
                    "Будь ласка, надішліть ваш емейл для реєстрації приєднання до календаря."
                );
                break;
            default:
                Utils.logInfo(`Unknown command: ${update.message.text}`);
                this.telegramBot.sendMessage(
                    update.message.chat.id,
                    "Невідома команда. Спробуйте /help."
                );
        }
    }

    handleCallbackQuery(update) {
        Utils.logInfo(`Handling callback query: ${update.data}`);

        if (update.data === 'cancel') {
            this.telegramBot.deleteMessage(update.message.chat.id, update.message.message_id);
            return;
        }

        const buttonData = JSON.parse(update.data);
        if (buttonData.type === "SelectActivity") {
            this.handleSelectActivity(update);
        } else if (buttonData.type === "SelectDate") {
            this.handleSelectDate(update);
        } else if (buttonData.type === "SelectTime") {
            this.handleSelectTime(update);
        } else if (buttonData.type === "Confirm") {
            this.handleConfirmTimeSlotsButton(update);
        }
        else {
            Utils.logInfo(`Handling unknown callback query data: ${update.data}`);
        }
    }

    handleConfirmTimeSlotsButton(update) {
        var data = JSON.parse(update.data);
        var date = data.date;
        Utils.logInfo(`Confirmed booking for date: ${date}`);
        var replyKeyboard = update.message.reply_markup.inline_keyboard;
        var selectedTimeSlots = [];
        replyKeyboard.forEach(row => {
            row.forEach(button => {
                if (button.text.includes('✅')) {
                    var selectedButton = JSON.parse(button.callback_data);
                    selectedTimeSlots.push(selectedButton.start + '-' + selectedButton.end);
                }
            });
        });

        var mergedTimeSlots = Utils.mergeTimeSlots(date, selectedTimeSlots);
        var timeSlotsText = mergedTimeSlots.map(slot => {
            return this.getTimeSlotTimeString(slot.from, slot.to);
        });
        timeSlotsText = mergedTimeSlots.length > 0 ? `\r\n${timeSlotsText.join('\r\n')}` : 'Час не обрано';
        Utils.logInfo(`Selected time slots: ${timeSlotsText}`);

        this.telegramBot.sendMessage(
            update.message.chat.id,
            `Ви забронювали на ${date}: ${timeSlotsText} \nДякуємо за бронювання!`
        );

        this.telegramBot.deleteMessage(update.message.chat.id, update.message.message_id);
    }

    handleSelectTime(update) {
        var replyKeyboard = update.message.reply_markup.inline_keyboard;
        replyKeyboard.forEach(row => {
            row.forEach(button => {
                if (button.callback_data === update.data) {
                    if (!button.text.includes('✅')) {
                        button.text += '✅'; // Add checkmark to the selected button
                    } else {
                        button.text = button.text.replace('✅', ''); // Remove any existing checkmark
                    }
                }
            });
        });

        this.telegramBot.editMessageReplyMarkup(
            update.message.chat.id,
            update.message.message_id,
            { inline_keyboard: replyKeyboard }
        );
    }

    handleSelectDate(update) {
        var data = JSON.parse(update.data);
        const date = data.date;
        Utils.logInfo(`Booking date: ${date}`);
        this.telegramBot.sendMessage(
            update.message.chat.id,
            `Ви обрали дату: ${date}. Будь ласка, оберіть час.`,
            { inline_keyboard: this.getDayTimeSlotsButtons(data) }
        );
    }

    getAvailableDates(parentData) {
        let availableDates = [];
        const today = new Date();
        for (let i = 0; i < 7; i++) {
            const date = new Date(today);
            date.setDate(today.getDate() + i);
            var buttonDate = date.toISOString().split('T')[0]; // Format date as YYYY-MM-DD
            var dayName = Utils.getUkrainianDayOfWeek(date);
            availableDates.push([{ text: buttonDate + " (" + dayName + ")", callback_data: ButtonData.fromDate(parentData, date).toString() }]);
        }
        availableDates.push([{ text: "Скасувати", callback_data: "cancel" }]);
        return availableDates;
    }

    getAvailableActivities(parentData) {
        return [
            { text: "Футбол", callback_data: ButtonData.fromActivity(parentData, 'football').toString() },
            { text: "Баскетбол", callback_data: ButtonData.fromActivity(parentData, 'basketball').toString() },
            { text: "Теніс", callback_data: ButtonData.fromActivity(parentData, 'tennis').toString() },
            { text: "Волейбол", callback_data: ButtonData.fromActivity(parentData, 'volleyball').toString() },
            { text: "Бадмінтон", callback_data: ButtonData.fromActivity(parentData, 'badminton').toString() },
            { text: "Інше", callback_data: ButtonData.fromActivity(parentData, 'other').toString() },
            { text: "Скасувати", callback_data: "cancel" }
        ];
    }

    getDayTimeSlotsButtons(parentData) {
        Utils.logInfo(`Getting time slots for date: ${parentData?.date}`);
        const timeSlots = [];
        const startTime = new Date(parentData?.date + 'T09:00:00'); // Start at 9 AM
        const endTime = new Date(parentData?.date + 'T20:00:00');
        var current = startTime;
        while (current < endTime) {
            var nextDate = Utils.dateAdd(current, 'minute', 30);
            var view = this.getTimeSlotTimeString(current, nextDate);
            var data = this.getTimeSlotButtonData(parentData, current, nextDate);
            var serializedData = data.toString();

            // TODO remove
            Utils.logInfo(`Prepared data: ${serializedData}`);

            timeSlots.push([{ text: view, callback_data: serializedData }]);
            current = nextDate;
        }
        timeSlots.push([{ text: "Підтвердити", callback_data: new ConfirmButtonData(parentData).toString() }]);
        timeSlots.push([{ text: "Скасувати", callback_data: "cancel" }]);
        return timeSlots;
    }

    getTimeSlotTimeString(from, to) {
        return `${from.toLocaleTimeString('uk-UA').slice(0, 5)} - ${to.toLocaleTimeString('uk-UA').slice(0, 5)}`;
    }

    getTimeSlotButtonData(parentData, from, to) {
        return ButtonData.fromTime(parentData, from, to);
    }

}

File: src/Telegram/PrivateChatManager.cs
using System.Text.RegularExpressions;
using Microsoft.Extensions.Caching.Memory;
using Newtonsoft.Json;
using Telegram.Bot;
using Telegram.Bot.Types;
using Telegram.Bot.Types.Enums;
using Telegram.Bot.Types.ReplyMarkups;
using TelegramBot.Calendar;
using TelegramBot.Configuration;
using TelegramBot.Core;
using TelegramBot.Exceptions;
using TelegramBot.Extensions;

namespace TelegramBot.Telegram;

public class PrivateChatManager
{
    private const string _checkMark = "✅";

    private static readonly Regex EmailRegex =
        new(
            @"^(([^<>()[\]\\,;:\s@']+(\.[^<>()[\]\\,;:\s@']\s)*)|.('.+'))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$");

    private readonly CalendarManager _calendarManager; // CalendarManager is a mock class
    private readonly MessageChainManager _chainManager;
    private readonly IConfigManager _configManager;
    private readonly ILogger<PrivateChatManager> _logger;
    private readonly IMemoryCache _memoryCache;
    private readonly ITelegramBotClient _telegramBot;
    private readonly MainGroupChat _mainGroupChat;
    private int _pullOffset;

    public PrivateChatManager(
        IConfigManager configManager,
        CalendarManager calendarManager,
        ITelegramBotClient telegramBot,
        MainGroupChat mainGroupChat,
        ILogger<PrivateChatManager> logger,
        IMemoryCache memoryCache)
    {
        _configManager = configManager;
        _calendarManager = calendarManager;
        _telegramBot = telegramBot;
        _mainGroupChat = mainGroupChat;
        _logger = logger;
        _memoryCache = memoryCache;
        _chainManager = new MessageChainManager(_memoryCache);
        int.TryParse(_configManager.GetProperty("PULL_OFFSET"), out _pullOffset);
    }

    public async Task SetupMenu()
    {
        var commands = new List<BotCommand>
        {
            new("register", "Зареєструватися"),
            new("book", "Забронювати"),
            new("cancel", "Скасувати бронь"),
            new("help", "Допомога")
        };
        await _telegramBot.SetMyCommands(commands, BotCommandScope.AllPrivateChats());
    }

    public async Task PollUpdates(int timeoutSeconds = 5)
    {
        try
        {
            var updates = await _telegramBot.GetUpdates(_pullOffset, 100, timeoutSeconds);

            if (updates.Length == 0)
            {
                _logger.LogInformation("No updates received or error occurred.");
                return;
            }

            await ProcessUpdates(updates);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing updates");
        }
    }

    private async Task ProcessUpdates(Update[] updates)
    {
        foreach (var update in updates) await ProcessUpdate(update);
        _pullOffset = updates.Last().Id + 1;
        _configManager.SetProperty("PULL_OFFSET", _pullOffset.ToString());
    }

    public async Task ProcessUpdate(Update update)
    {
        try
        {
            var updateLogData = JsonConvert.SerializeObject(update, JsonSetting.Intended);
            var message = update?.Message ?? update?.CallbackQuery?.Message;
            var chatType = message?.Chat?.Type;
            if (chatType != ChatType.Private) return;

            // TODO Add validations for user update 
            if (!await ValidateUserAsync(message)) 
                return;

            _logger.LogInformation($"Received update {updateLogData}");

            if (update.Message != null)
            {
                if (string.IsNullOrEmpty(message.Text))
                {
                    _logger.LogInformation($"Received message without text from {message.From?.Username}");
                    return;
                }

                _logger.LogInformation($"Received message: {message.Text} from {message.From?.Username}");
                if (message.Text.StartsWith("/"))
                    await HandleCommand(update);
                else if (message.Text.Contains("@"))
                    await HandleEmailMessage(message);
            }
            else if (update.CallbackQuery != null)
            {
                var callbackQuery = update.CallbackQuery;
                _logger.LogInformation(
                    $"Received callback query: {callbackQuery.Data} from {callbackQuery.From?.Username}");
                await HandleCallbackQuery(callbackQuery);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing update update: {Update}",
                JsonConvert.SerializeObject(update, JsonSetting.Intended));
            await NotifyAdmins(ex);
            await SendMessageToUser(update);
        }
    }

    private async Task<bool> ValidateUserAsync(Message message)
    {
        if (message.From?.IsBot is true && message.From.Username != _configManager.GetBotName())
            return false;
        
        // 1) No username
        if (message.Chat.Username is null)
        {
            await _telegramBot.SendMessage(message.Chat.Id,
                "Щоб забронювати майданчик, треба створити ім'я користувача\r\n" +
                "https://www.youtube.com/watch?v=Q-iZWJ7IwZs");
            return false;
        }
        
        // 2) Not in group channel
        var isInGroupChat = await _mainGroupChat.IsInGroupChat(message.From.Id);
        if (!isInGroupChat)
        {
            await _telegramBot.SendMessage(message.Chat.Id,
                "Вас нема в чаті спортмайданчику або вас забанили");
            return false;
        }
        
        // 3) TODO Banned
        
        
        return true;
    }

    private async Task SendMessageToUser(Update update)
    {
        try
        {
            var message = update?.Message ?? update?.CallbackQuery?.Message;
            if (message == null)
                return;

            await _telegramBot.SendMessage(message.Chat.Id, "Сталася помилка, вибачте за незручності");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error during user error notification");
        }
    }

    private async Task NotifyAdmins(Exception exception)
    {
        try
        {
            var adminsChat = _configManager.GetAdminsChats();
            foreach (var chatId in adminsChat) await _telegramBot.SendMessage(chatId, $"Error: {exception.Message}");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error during admins notification");
        }
    }

    private async Task HandleCommand(Update update)
    {
        Message createdMessage = null;
        switch (update.Message.Text)
        {
            case "/book":
                _logger.LogInformation("Handling book command");
                createdMessage = await _telegramBot.SendMessage(
                    update.Message.Chat.Id,
                    "Оберіть активність:",
                    replyMarkup: new InlineKeyboardMarkup(GetAvailableActivities(null))
                );
                break;
            case "/help":
            case "/start":
                _logger.LogInformation("Handling help command");
                createdMessage = await _telegramBot.SendMessage(
                    update.Message.Chat.Id,
                    "Довідка: використовуйте /book для бронювання.\nДля реєстрації/оновлення емейла використовуйте /register."
                );
                break;
            case "/register":
                _logger.LogInformation("Handling register command");
                createdMessage = await _telegramBot.SendMessage(
                    update.Message.Chat.Id,
                    "Будь ласка, надішліть ваш емейл для реєстрації приєднання до календаря."
                );
                break;
            case "/cancel":
                await CancelBooking(update);
                break;
            default:
                _logger.LogInformation($"Unknown command: {update.Message.Text}");
                createdMessage = await _telegramBot.SendMessage(
                    update.Message.Chat.Id,
                    "Невідома команда. Спробуйте /help."
                );
                break;
        }

        if (createdMessage != null)
            _chainManager.CreateChain(createdMessage.MessageId, update.Message.MessageId);
    }

    private async Task HandleEmailMessage(Message message)
    {
        if (!EmailRegex.IsMatch(message.Text))
        {
            await _telegramBot.SendMessage(message.Chat.Id,
                $"Не вірно введений емейл {message.Text}");
        }
        else
        {
            await _telegramBot.UnpinAllChatMessages(message.Chat.Id);

            _logger.LogInformation($"Email detected: {message.Text}");
            _memoryCache.CreateEntry(message.Chat.Username).Value = message.Text;

            await _telegramBot.PinChatMessage(message.Chat.Id, message.MessageId);
            await _telegramBot.SendMessage(message.Chat.Id,
                $"Ваш емейл {message.Text} успішно зареєстровано. Дякуємо!");
        }
    }

    private async Task CancelBooking(Update update)
    {
        _logger.LogInformation("Handling cancel command");
        var userName = update.Message.Chat.Username;
        var buttons = new List<List<InlineKeyboardButton>>();
        await foreach (var userTimeSlot in _calendarManager.GetUserBookingAsync(userName))
        {
            var day = userTimeSlot.Start.Date.ToString("dd-MM-yy");
            var time = $"{userTimeSlot.Start:hh:mm}-{userTimeSlot.End:hh:mm}";
            var text = $"{userTimeSlot.Summary} {day} {time}";
            var callbackDataOrUrl = EventButtonData.FromEventId(userTimeSlot.EventId).ToJson();
            buttons.Add([new InlineKeyboardButton(text, callbackDataOrUrl)]);
        }

        if (buttons.Count == 0)
        {
            await _telegramBot.SendMessage(
                update.Message.Chat.Id,
                "Нема чого скасовувати"
            );
            return;
        }

        buttons.Add([new InlineKeyboardButton("Підтвердити скасування", new ConfirmCancelButtonData().ToJson())]);
        buttons.Add([new InlineKeyboardButton("Не зараз", "cancel")]);

        var createdMessage = await _telegramBot.SendMessage(
            update.Message.Chat.Id,
            "Будь ласка, Оберіть слоти які ви хочете скасувати:.", replyMarkup: new InlineKeyboardMarkup(buttons)
        );

        _chainManager.CreateChain(createdMessage.MessageId, update.Message.MessageId);
    }

    private async Task HandleCallbackQuery(CallbackQuery update)
    {
        _logger.LogInformation($"Handling callback query: {update.Data}");

        if (update.Data == "cancel")
        {
            await DeleteMessage(update.Message.Chat.Id, update.Message.MessageId);
            return;
        }

        var buttonData = JsonConvert.DeserializeObject<ButtonData>(update.Data);

        switch (buttonData.Type)
        {
            case "SelectActivity":
                await HandleActivitySelection(update);
                break;
            case "SelectDate":
                await HandleSelectDate(update);
                break;
            case "SelectTime":
                await HandleSelectTime(update);
                break;
            case "Confirm":
                await HandleConfirmTimeSlotsButton(update);
                break;
            case "ConfirmCancel":
                await HandleConfirmCancelButton(update);
                break;
            case "SelectEvent":
                await HandleSelectTime(update);
                break;
            default:
                _logger.LogInformation($"Handling unknown callback query data: {update.Data}");
                break;
        }
    }

    private async Task HandleActivitySelection(CallbackQuery update)
    {
        var buttonData = JsonConvert.DeserializeObject<ButtonData>(update.Data);
        var message = await _telegramBot.SendMessage(
            update.Message.Chat.Id,
            "Оберіть дату та час для бронювання:",
            replyMarkup: new InlineKeyboardMarkup { InlineKeyboard = await GetAvailableDates(buttonData) }
        );
        _chainManager.CreateChain(message.MessageId, update.Message.MessageId);
    }

    private async Task HandleConfirmTimeSlotsButton(CallbackQuery update)
    {
        var data = JsonConvert.DeserializeObject<ButtonData>(update.Data);
        var replyKeyboard = update.Message.ReplyMarkup;
        var selectedTimeSlots = new List<string>();
        var timeSlots = new List<TimeSlot>();

        foreach (var row in replyKeyboard.InlineKeyboard)
        foreach (var button in row)
            if (button.Text.Contains(_checkMark))
            {
                var selectedButton = JsonConvert.DeserializeObject<ButtonData>(button.CallbackData);
                selectedTimeSlots.Add($"{selectedButton.Start}-{selectedButton.End}");
                timeSlots.Add(new TimeSlot
                    { Start = TimeOnly.Parse(selectedButton.Start), End = TimeOnly.Parse(selectedButton.End) });
            }

        var mergedTimeSlots = Utils.MergeTimeSlots(selectedTimeSlots);
        var timeSlotsText = mergedTimeSlots.Count > 0
            ? string.Join("\r\n", mergedTimeSlots.Select(slot => slot.ToView()))
            : "Час не обрано";

        _logger.LogInformation($"Selected time slots: {timeSlotsText}");

        var activity = Activities.All[data.Activity] ?? "Інше";

        var chat = update.Message.Chat;
        if (mergedTimeSlots.Any())
        {
            if (!_memoryCache.TryGetValue(chat.Username, out string email))
            {
                var chatInfo = await _telegramBot.GetChat(chat.Id);
                email = chatInfo.PinnedMessage?.Text;
            }

            var date = DateOnly.Parse(data.Date);
            
            try
            {
                await _calendarManager.BookSlots(chat.Username, email, activity, date, timeSlots);
            }
            catch (CalendarBookingRaceConditionException)
            {
                await RebuildTimeSlotsKeyboard(update.Message.Chat.Id, update.Message.MessageId, data.Date, data.Activity);
                await _telegramBot.SendMessage(chat.Id, $"Нажаль, вже хтось забронював цей час швидше. Спробуйте ще");
                return;
            }
            
            await _telegramBot.SendMessage(
                chat.Id,
                $"Ви забронювали на {data.Date} під {activity}: {timeSlotsText} \nДякуємо за бронювання!");
        }
        else
        {
            await _telegramBot.SendMessage(
                chat.Id,
                $"Нажаль вже все заброньовано {data.Date}");
        }

        await DeleteMessage(chat.Id, update.Message.MessageId);
    }

    private async Task HandleConfirmCancelButton(CallbackQuery update)
    {
        var replyKeyboard = update.Message.ReplyMarkup;
        var selectedEvents = new List<string>();
        var selectedButtonsTexts = new List<string>();
        
        foreach (var row in replyKeyboard.InlineKeyboard)
        foreach (var button in row)
            if (button.Text.Contains(_checkMark))
            {
                var selectedButton = JsonConvert.DeserializeObject<EventButtonData>(button.CallbackData);
                selectedEvents.Add(selectedButton.EventId);
                selectedButtonsTexts.Add(button.Text.Replace(_checkMark, ""));
            }

        await _calendarManager.CancelEvents(selectedEvents);

        await _mainGroupChat.NotifyAsync($"Cкасовано бронювання:\r\n{string.Join("\r\n",selectedButtonsTexts)}");

        await DeleteMessage(update.Message.Chat.Id, update.Message.MessageId);
    }

    private async Task DeleteMessage(long chatId, int messageId)
    {
        var chain = _chainManager.GetChain(messageId).Distinct();
        await chain.Foreach(async chainMessageId =>
        {
            try
            {
                await _telegramBot.DeleteMessage(chatId, chainMessageId);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error deleting message");
            }
            finally
            {
                _chainManager.Remove(chainMessageId);
            }
        });
    }

    private async Task HandleSelectTime(CallbackQuery update)
    {
        var replyKeyboard = update.Message.ReplyMarkup;

        foreach (var row in replyKeyboard.InlineKeyboard)
        foreach (var button in row)
            if (button.CallbackData == update.Data)
            {
                if (!button.Text.Contains(_checkMark))
                    button.Text = _checkMark + button.Text;
                else
                    button.Text = button.Text.Replace(_checkMark, "");
            }

        await _telegramBot.EditMessageReplyMarkup(
            update.Message.Chat.Id,
            update.Message.MessageId,
            new InlineKeyboardMarkup { InlineKeyboard = replyKeyboard.InlineKeyboard }
        );
    }

    private async Task HandleSelectDate(CallbackQuery update)
    {
        var data = JsonConvert.DeserializeObject<ButtonData>(update.Data);
        var date = data.Date;
        var activity = data.Activity;
        _logger.LogInformation($"Booking date: {date}");
        var createdMessage = await _telegramBot.SendMessage(
            update.Message.Chat.Id,
            $"Ви обрали дату: {date}. Будь ласка, оберіть час.",
            replyMarkup: new InlineKeyboardMarkup { InlineKeyboard = await GetDayTimeSlotsButtons(date, activity) }
        );

        _chainManager.CreateChain(createdMessage.MessageId, update.Message.MessageId);
    }

    private async Task<List<List<InlineKeyboardButton>>> GetAvailableDates(ButtonData parentData)
    {
        var buttons = new List<List<InlineKeyboardButton>>();
        var calendar = await _calendarManager.GetCalendar();
        var availableDays = calendar.GetDaysWithEmptySlots();
        foreach (var day in availableDays)
        {
            var buttonDate = day.ToString("yyyy-MM-dd");
            var dayName = Utils.GetUkrainianDayOfWeek(day.DayOfWeek);
            buttons.Add([
                new InlineKeyboardButton
                {
                    Text = $"{buttonDate} ({dayName})", CallbackData = ButtonData.FromDate(parentData, day).ToJson()
                }
            ]);
        }

        buttons.Add([new InlineKeyboardButton { Text = "Скасувати", CallbackData = "cancel" }]);
        return buttons;
    }

    private List<List<InlineKeyboardButton>> GetAvailableActivities(ButtonData parentData)
    {
        return Activities.All.Select((t, i) => (List<InlineKeyboardButton>)
                         [
                             new InlineKeyboardButton
                                 { Text = t, CallbackData = ButtonData.FromActivity(parentData, i).ToJson() }
                         ])
                         .ToList();
    }

    
    private async Task RebuildTimeSlotsKeyboard(long chatId, int messageId, string dateView, int activity)
    {
        var inlineKeyboardMarkup = new InlineKeyboardMarkup { InlineKeyboard = await GetDayTimeSlotsButtons(dateView, activity) };
        await _telegramBot.EditMessageReplyMarkup(chatId, messageId, inlineKeyboardMarkup);
    }
    
    private async Task<List<List<InlineKeyboardButton>>> GetDayTimeSlotsButtons(string dateView, int activity)
    {
        _logger.LogInformation($"Getting time slots for date: {dateView}");

        var date = DateTime.Parse(dateView);
        var buttons = new List<List<InlineKeyboardButton>>();
        var calendar = await _calendarManager.GetCalendar();
        var emptySlots = calendar.GetEmptySlots(DateOnly.FromDateTime(date));

        foreach (var emptySlot in emptySlots)
            buttons.Add([
                new InlineKeyboardButton(emptySlot.ToView(),
                    GetTimeSlotButtonData(emptySlot.Start, emptySlot.End).ToJson())
            ]);

        buttons.Add([
            new InlineKeyboardButton { Text = "Підтвердити", CallbackData = new ConfirmButtonData(dateView, activity).ToJson() }
        ]);
        buttons.Add([new InlineKeyboardButton { Text = "Скасувати", CallbackData = "cancel" }]);
        return buttons;
    }

    private ButtonData GetTimeSlotButtonData(TimeOnly from, TimeOnly to)
    {
        return ButtonData.FromTime(from, to);
    }

    public async Task SetupWebhook(string url)
    {
        var allowedUpdates = new[] { UpdateType.Message, UpdateType.CallbackQuery, UpdateType.InlineQuery };
        var webhookSecret = _configManager.GetTelegramWebhookSecret();
        await _telegramBot.SetWebhook(url, allowedUpdates: allowedUpdates, secretToken: webhookSecret);
    }
}